### DISCUSSION TITLE
バックグランド時に音声のみを配信する方法について

### DISCUSSION THREAD

**Discussion by Rikukuku** [HaishinKit/HaishinKit.swift#1011] [Q&A]
## 概要
配信中のアプリをバックグランドにしたときに音声のみを配信することは可能でしょうか？

過去のIssuueを拝見させてもらったところ、映像はバックグランドでは配信不可能ということがわかりましたが、
[音声であれば配信可能](https://github.com/shogo4405/HaishinKit.swift/issues/124#issuecomment-253911285)であるという旨のコメントを確認したので、お伺いしました。

もし可能であれば、実装方法などをご教授いただけますと幸いです。

## 確認したIssue
https://github.com/shogo4405/HaishinKit.swift/issues/124
https://github.com/shogo4405/HaishinKit.swift/issues/525
https://github.com/shogo4405/HaishinKit.swift/issues/626
https://github.com/shogo4405/HaishinKit.swift/issues/847

**Comment by shogo4405** [comment]
次の設定でできると思いますよ。
- [Project設定] → [Signing & Capabilities] → [Background Modes]より、Audio, AirPlay and Picture In Picture をON

**Comment by Rikukuku** [comment]
ご返信ありがとうございます。

> [Project設定] → [Signing & Capabilities] → [Background Modes]より、Audio, AirPlay and Picture In Picture をON
上記を追加して実機でテストすると一瞬だけマイク利用時の赤い帯が画面上部に現れるようになりましたが、すぐに消えてバックグランド処理が消えているようです。

AppDelegateには以下の処理を追加しているのですが、その他必要なコードや設定などはありますでしょうか？

```
        let session = AVAudioSession.sharedInstance()
        do {
            if #available(iOS 10.0, *) {
                try session.setCategory(.playAndRecord, mode: .default, options: [.defaultToSpeaker, .allowBluetooth])
            } else {
                session.perform(NSSelectorFromString("setCategory:withOptions:error:"), with: AVAudioSession.Category.playAndRecord, with: [
                    AVAudioSession.CategoryOptions.allowBluetooth,
                    AVAudioSession.CategoryOptions.defaultToSpeaker
                ])
                try session.setMode(.default)
            }
            try session.setActive(true)
        } catch {
            logger.error(error)
        }
```

## 環境
xcode: 12.5.1
swift:  5.4.2
iphone: 8
ios: 14.8

**Comment by shogo4405** [comment]
- iOS12 の時代だと既存のコードでできた記憶があります。
- iOS15 の時代だと、カメラを利用している場合にバックグラウンドにうつると`AVCaptureSession`が停止させられているようでした
   - 回避方法として、バックグラウンドとフォアグランド時にカメラのデタッチとアタッチを行えば動きました
``` 
        NotificationCenter.default.addObserver(self, selector: #selector(didEnterBackground(_:)), name: UIApplication.didEnterBackgroundNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(didBecomeActive(_:)), name: UIApplication.didBecomeActiveNotification, object: nil)

   @objc
    private func didEnterBackground(_ notification: Notification) {
        rtmpStream.attachCamera(nil)
    }

    @objc
    private func didBecomeActive(_ notification: Notification) {
        rtmpStream.attachCamera(DeviceUtil.device(withPosition: currentPosition)) { error in
            logger.warn(error.description)
        }
    }
```

**Comment by Rikukuku** [comment]
iOSのバージョンの問題だったのですね...
ご教授頂いたコードを追加すると、正常に動作するようになりました！

お忙しい中ご返信ありがとうございました。

**Comment by jexwang** [comment]
Hello,

At ver 1.4.0, this framework supported background audio stream, but I can't get this function to work at ver 1.4.4.

And I try to set breakpoint in "IOMixer.didEnterBackground(_ notification:)" this function, it's will not trigger when app enter background.

Device: iPhone X
OS: 15.7.1

Can you help me? Thanks!

**Reply by shogo4405** [reply]
Right. I fixed https://github.com/shogo4405/HaishinKit.swift/pull/1161. I will test then release in a few days.

**Reply by shogo4405** [reply]
I saw release 1.4.0. It's it seems don't work background audio stream. 
I changed to release plan 1.4.5 to 1.5.0.

**Reply by luynv-2141** [reply]
@shogo4405 hello, I tried version 1.4.5 and audio stream still not work when app was entered in background. I tried on code example and it not work too. Please give me some supports.

Device: iPhone XS Max
OS: 16.3.1
